### ISSUE TITLE
Whisper large v3 model repeats a lot

### ISSUE THREAD

**Issue by sindresorhus** [ggml-org/whisper.cpp#1507]
I have gotten many reports that the large v3 model repeats sentences much more often than v2. I'm not sure if there's anything Whisper.cpp can do about this.

**Comment by ggerganov** [comment]
If we can determine in some way that the problem is `whisper.cpp` related, then I'll look into it more.
But so far, my analysis indicate that the problem is in the v3 model itself as I observe similar issues with the OpenAI implementation.

**Comment by sindresorhus** [comment]
Yeah, the problem seems to be the model itself: https://github.com/openai/whisper/discussions/1762#discussioncomment-7532295

The problem seems to occur during silence, so maybe Whisper.cpp could remove silence from audio?

**Comment by ggerganov** [comment]
Removing silence from the audio is outside the scope of `whisper.cpp` because AFAIK there are many different algorithms to achieve this. It's better to leave it to the 3rd party to decide which one to use for their specific case.

**Comment by sindresorhus** [comment]
Repetition on silences is a big problem with Whisper in general, large v3 just made it even worse. Having it built into Whisper.cpp would improve the general Whisper quality for all consumers, instead of every consumer having to implement a custom solution. It could be opt-in. I think it's worth considering.

In my case, I haven't found any good solutions for it that is not Python based. I need it to be C++/C/Swift.

**Comment by ggerganov** [comment]
We can add a naive VAD as an optional pre-processing step, but I'm doubtful that it will help much, because the samples that I see failing with v3 do not contain silences.

Here are some strategies that I've observed to reduce repetition and hallucinations:
- Use 5 beams
- Increase entropy threshold from the default `2.4` to `2.8` for example. Higher threshold will reject repetitive text and fallback to sampling with higher temperature
- Reduce the maximum context size (`--max-context`). By default it is `224`. Setting it to `64` or `32` can reduce the repetitions significantly. Setting it to `0` will most likely eliminate all repetitions, but the transcription quality can be affected because it will be losing the context from the previous transcript

**Comment by bobqianic** [comment]
> Repetition on silences is a big problem with Whisper in general

I think it's possible for someone to fine-tune the model using `silence audio`. Even the largest whisper model, which has **`1.5 B`** parameters, is considered relatively small in the context of today's LLM : )

**Comment by dubefab** [comment]
> We can add a naive VAD as an optional pre-processing step, but I'm doubtful that it will help much, because the samples that I see failing with v3 do not contain silences.
> 
> Here are some strategies that I've observed to reduce repetition and hallucinations:
> 
> * Use 5 beams
> * Increase entropy threshold from the default `2.4` to `2.8` for example. Higher threshold will reject repetitive text and fallback to sampling with higher temperature
> * Reduce the maximum context size (`--max-context`). By default it is `224`. Setting it to `64` or `32` can reduce the repetitions significantly. Setting it to `0` will most likely eliminate all repetitions, but the transcription quality can be affected because it will be losing the context from the previous transcript

I tried this with large-v2 and it made it even better!

**Comment by jxy** [comment]
There is the no speech token that currently whisper.cpp ignores

https://github.com/ggerganov/whisper.cpp/blob/447d49530c9af41fe24f2ae510f452903dba330d/whisper.cpp#L4592

Actually implement no speech threshold similar to openai/whisper might help.

**Comment by ex3ndr** [comment]
I am trying to work-around this problem and VAD is not useful since if there even small silence interval it would emit something. whisper.cpp ignores "no speak" token while it is very crucial for it to work and it seems impossible to make it work without it.

**Comment by ex3ndr** [comment]
I have added a PR to return nosp token: https://github.com/ggerganov/whisper.cpp/pull/1588

**Comment by itsthisjustin** [comment]
> * Reduce the maximum context size (`--max-context`). By default it is `224`. Setting it to `64` or `32` can reduce the repetitions significantly. Setting it to `0` will most likely eliminate all repetitions, but the transcription quality can be affected because it will be losing the context from the previous transcript

I don't see this as a param I can use in the Swift Package. What am I missing?

**Comment by aiyinyuedejustin** [comment]
same thing here... 
<img width="613" alt="image" src="https://github.com/user-attachments/assets/8b8415c3-c277-4a0b-b77d-86b27d330add" />

**Comment by gregsadetsky** [comment]
Just to reiterate a big thanks to everyone & that passing the options below (gleaned from past comments here) did solve the loops/hallucination on my end!

The flags I now pass to `whisper-cli` are:

`
whisper-cli -nt --entropy-thold 2.8 --max-context 64 -vm /.../models/ggml-silero-v5.1.2.bin --vad -m /.../models/ggml-large-v2.bin -f (file)
`

Works great for me - thanks again!

**Comment by LaurinmyReha** [comment]
maybe this is useful for someone here still:
I wrote a article about this
https://nyra-labs.com/research/killing-hallucinations
which improves this behaviour quite drastically. The rewind and escape decoding trick could also just be used for any whisper model really without retraining. 
Feel free to try out the model too :)
https://huggingface.co/nyralabs/CrisperWhisper2.0_large
